src/main/scala/cosas/fns.scala

```diff
@@ -47,4 +47,156 @@ object fns {
 
   trait Fn3[A, B, C] extends AnyFn3 { type In1 = A; type In2 = B; type In3 = C }
 
+
+
+  /* Dependent functions aka dependent products */
+  trait AnyDepFn { type Out }
+  trait AnyDepFn0 extends AnyDepFn
+  trait AnyDepFn1 extends AnyDepFn { type In1 }
+  trait DepFn1[I,O] extends AnyDepFn1 { type In1 = I; type Out = O }
+  trait AnyDepFn2 extends AnyDepFn { type In1; type In2 }
+  trait DepFn2[I1,I2,O] extends AnyDepFn2 { type In1 = I1; type In2 = I2; type Out = O }
+  // TODO moar arities
+  // trait AnyDepFn3 extends AnyDepFn { type In1; type In2 }
+
+  trait AnyDepFn1Composition extends AnyDepFn1 {
+
+    type First <: AnyDepFn1
+    val first: First
+    type Second <: AnyDepFn1 { type In1 >: First#Out }
```

Comment:
I think it's better to do it another way round: `First <: AnyDepFn1 { type Out <: Second#In1 }`, or require them equal in `Second`